search/custom-embeddings/custom_embeddings.ipynb

```diff
@@ -0,0 +1,800 @@
+    "            ]\n",
+    "        )\n",
+    "        embs.extend([e.values for e in result])\n",
+    "    return embs"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Get embeddings for the question titles/body and add them as the `\"embedding\"` column."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "df[\"title_body\"] = df[\"title\"] + \"\\n\" + df[\"body\"]\n",
+    "\n",
+    "df = df.assign(embedding=get_embeddings_wrapper(df.title_body))\n",
+    "df.head()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "TKBmi2BMk_OU"
+   },
+   "source": [
+    "## Scrape HTML from Question Pages\n",
+    "\n",
+    "- Grab HTML to upload to Cloud Storage\n",
+    "- This will be used for the search results links"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "tXHmC10IitET"
+   },
+   "outputs": [],
+   "source": [
+    "JSONL_MIME_TYPE = \"application/jsonl\"\n",
+    "HTML_MIME_TYPE = \"text/html\"\n",
+    "\n",
+    "BUCKET_NAME = \"ucs-demo\"\n",
+    "DIRECTORY = \"embeddings-stackoverflow\"\n",
+    "BLOB_PREFIX = f\"{DIRECTORY}/html/\"\n",
+    "\n",
+    "GCS_URI_PREFIX = f\"gs://{BUCKET_NAME}/{BLOB_PREFIX}\"\n",
+    "\n",
+    "from google.cloud import storage\n",
+    "\n",
+    "storage_client = storage.Client()\n",
+    "bucket = storage_client.bucket(BUCKET_NAME)\n",
+    "\n",
+    "\n",
+    "def scrape_question(question_url: str) -> str:\n",
+    "    response = requests.get(question_url)\n",
+    "\n",
+    "    if response.status_code != 200 or not response.content:\n",
+    "        print(f\"URL: {question_url} Code: {response.status_code}\")\n",
+    "        return None\n",
+    "\n",
+    "    print(f\"Scraping {question_url}\")\n",
+    "\n",
+    "    link_title = response.url.split(\"/\")[-1] + \".html\"\n",
+    "    gcs_uri = f\"{GCS_URI_PREFIX}{link_title}\"\n",
+    "\n",
+    "    # Upload HTML to Google Cloud Storage\n",
+    "    blob = bucket.blob(f\"{BLOB_PREFIX}{link_title}\")\n",
+    "    blob.upload_from_string(response.content, content_type=HTML_MIME_TYPE)\n",
+    "    time.sleep(1)\n",
+    "    return gcs_uri"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Get the published URL from the ID\n",
+    "QUESTION_BASE_URL = \"https://stackoverflow.com/questions/\"\n",
+    "df[\"question_url\"] = df[\"id\"].apply(lambda x: f\"{QUESTION_BASE_URL}{x}\")\n",
+    "\n",
+    "# Scrape HTML from stackoverflow.com and upload to GCS\n",
+    "df[\"gcs_uri\"] = df[\"question_url\"].apply(scrape_question)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Restructure the embeddings data to follow Vertex AI Search format (Unstructured with Metadata)"
```

Comment:
This is probably a good place to link to the different ways of ingesting this into Vertex AI Search and explain why you chose unstructured w/ metadata.